torch/csrc/api/include/torch/nn/modules/any.h

```diff
@@ -552,5 +565,80 @@ ModuleType& AnyModule::get_(
       c10::demangle(typeid(ModuleType).name()));
 }
 
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ NamedAnyModule ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+class NamedAnyModule {
+ public:
+  /// We don't allow constructing an empty `NamedAnyModule`.
+  NamedAnyModule() = delete;
+
+  /// Creates a `NamedAnyModule` from a (boxed) `Module`.
+  template <typename ModuleType>
+  NamedAnyModule(std::string name, std::shared_ptr<ModuleType> module_ptr)
+      : NamedAnyModule(std::move(name), AnyModule(std::move(module_ptr))) {}
+
+  /// Creates a `NamedAnyModule` from a `Module`, moving or copying it
+  /// into a `shared_ptr` internally.
+  // NOTE: We need to use `std::remove_reference<M>::type` to get rid of
+  // any reference components for make_unique.
+  template <typename M, typename = torch::detail::enable_if_module_t<M>>
+  NamedAnyModule(std::string name, M&& module)
+      : NamedAnyModule(
+          std::move(name),
+          std::make_shared<typename std::remove_reference<M>::type>(
+            std::forward<M>(module))) {}
+
+  /// Creates a `NamedAnyModule` from a `Module` that is unwrapped from
+  /// a `ModuleHolder`.
+  template <typename M>
+  NamedAnyModule(std::string name, const ModuleHolder<M>& module_holder)
+      : NamedAnyModule(std::move(name), module_holder.ptr()) {}
+
+  /// Returns a reference to the name.
+  std::string name() const noexcept {
+    return name_;
+  }
+
+  /// Returns a reference to the module.
+  AnyModule& module() noexcept {
+    return module_;
+  }
+
+ private:
+  /// Creates a `NamedAnyModule` from a type-erased `AnyModule`.
+  NamedAnyModule(std::string name, AnyModule any_module) {
+    name_ = std::move(name);
+    module_ = std::move(any_module);
+  }
+
+  std::string name_;
+  AnyModule module_;
+};
+
+// NOTE: We might wonder why we need to have the `NamedAnyModule` class and have
+// the `modules_ordered_dict()` function signature be
+// `modules_ordered_dict(std::initializer_list<NamedAnyModule> named_modules)`,
+// instead of
+// `modules_ordered_dict(std::initializer_list<torch::OrderedDict<
+//    std::string, ModuleType>::Item> named_modules)`.
+// The reason is that when we pass in a braced-init list such as
+// `modules_ordered_dict({{"m1", M(1)}, {"m2", M(2)}})`,
+// if we use the second signature, at the template argument deduction step
+// the compiler is not able to deduce the type of `ModuleType` to the type of
+// `M(1)` or `M(2)`, since the compiler doesn't actually look into the
```

Comment:
Did you try whether `std::pair<std::string, AnyModule>` works?

Reply:
I tried to change the `modules_ordered_dict(...)` function to ```torch::OrderedDict<std::string, AnyModule> modules_ordered_dict(std::initializer_list<std::pair<std::string, AnyModule>> named_modules)```, however it doesn't seem to work and throws:
```cpp
../test/cpp/api/sequential.cpp: In member function ‘virtual void SequentialTest_ConstructsFromConcreteType_Test::TestBody()’:
../test/cpp/api/sequential.cpp:72:4: error: could not convert ‘{{"m1", SequentialTest_ConstructsFromConcreteType_Test::TestBody()::M(1)}, {std::__cxx11::basic_string<char>(((const char*)"m2"), std::allocator<char>()), SequentialTest_ConstructsFromConcreteType_Test::TestBody()::M(2)}, {"m3", SequentialTest_ConstructsFromConcreteType_Test::TestBody()::M(3)}}’ from ‘<brace-enclosed initializer list>’ to ‘std::initializer_list<std::pair<std::__cxx11::basic_string<char>, torch::nn::AnyModule> >’
```
The compiler is not able to match `std::initializer_list<std::pair<std::string, AnyModule>>` to the nested braced-init list `{{"m1", M(1)}, {std::string("m2"), M(2)}, {"m3", M(3)}}`. So I think the NamedAnyModule approach here is necessary.